### ISSUE TITLE
rsnapreport does not work with backup_script 

### ISSUE THREAD

**Issue by mbosecke** [rsnapshot/rsnapshot#52]
Using `backup_script` does not output anything useful when piped through rsnapreport.

My understanding is that rsnapreport relies on the statistics generated by rsync, however `backup_script` uses a custom "sync_if_different" method instead of using rsync. Therefore, there are no statistics and no report.

**Comment by djk20** [comment]
backup_script is intended for things like taking a dump of a database into
a dump directory, so that rsnapshot can then backup that dump directory (in
a later backup line) rather than rsyncing the internal database files
(which might be in an inconsistent state and might only work with a certain
version of the database software).

In that scenario, the backup_script is not doing rsync or anything like
rsync, it is a local operation in the sense that it does not involve a
network transfer of files.  I don't see how it would make sense for
rsnapreport to include a conversion of internal database files on host fred
to portable dump format files on host fred, in its statistics.

backup_script can also be used (or some might say misused) for things like
  backup_script    /bin/date "+ backup of example.com started at %c"
unused1
but then it makes even less sense to have rsnapreport trying to gather
statistics on how long it takes to run a date command.

In short, backup_script is not intended for making actual backups (eg with
rsync) - that's what the "backup" line is for.  I don't see how rsnapreport
could do stats for a backup_script or why you would want it to.



On 6 February 2015 at 04:06, Mitchell Bösecke <notifications@github.com>
wrote:

> Using backup_script does not output anything useful when piped through
> rsnapreport.
>
> My understanding is that rsnapreport relies on the statistics generated by
> rsync, however backup_script uses a custom "sync_if_different" method
> instead of using rsync. Therefore, there are no statistics and no report.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/DrHyde/rsnapshot/issues/52>.
>



-- 

David Keegel <david.keegel@gmail.com>

**Comment by mbosecke** [comment]
It is indeed a local operation, the output is dumped into a temp folder and then the contents of that temp folder is synced with the backup root. Because it's ultimately being synced with the backup root I would argue that it IS used for making actual backups. I think there is room for improvement in gathering statistics during that syncing phase of the temp folder with the backup root.

In our case, our backup script is responsible for remotely connecting to a postgres database and dumping out a backup. Something like this, perhaps labeled `backup-postgresql.sh`:

    ssh postgres@database-server 'pg_basebackup --format=tar --pgdata=-' > pgbackup.tar.gz

and then our rsnapshot config contains:

    backup_script backup-postgresql.sh ./

So our rsnapshot config file only contains one "backup_script" directive without a "backup" directive. Rsnapshot doesn't complain about this setup and it allows for us to use no unnecessary extra disk space on the `database-server` due to the fact that `pgbackup.tar.gz` is written over the ssh tunnel into a temp directory on the backup server.

**Comment by bebehei** [comment]
> The output of the backup_script is dumped into a temp folder and then the contents of that temp folder is synced with the backup root. Because it's ultimately being synced with the backup root I would argue that it IS used for making actual backups. I think there is room for improvement in gathering statistics during that syncing phase of the temp folder with the backup root.

There is no "sync" done when you use backup_script. A temporary folder is created, your command gets executed in this folder and after that, it will get copied with plain cp. rsync is not used for backup_script for this but that is what `rsnapreport.pl` depends on.

@djk20 pointed it out already. The rsync-output will be about the actual changes, which got transmitted. So if rsync found changes of x GBs, and y files, those will get displayed by rsnapreport.pl. If we would use rsync in your case, you will get statistics how fast it copied the tar-file back to your snapshot-root, but you won't get statistics how long the network-transfer needed, and so on.

Disclaimer: I don't use rsnapreport.pl. I only looked at the code right now.

> In our case, our backup script is responsible for remotely connecting to a postgres database and dumping out a backup. Something like this, perhaps labeled `backup-postgresql.sh`:

>    ssh postgres@database-server 'pg_basebackup --format=tar --pgdata=-' > pgbackup.tar.gz

> and then our rsnapshot config contains:

>    backup_script backup-postgresql.sh ./

In your case, I would just use a normal shell-script which is executed by a cronjob and looks a bit like this.

    ssh postgres@database-server 'pg_basebackup --format=tar --pgdata=-' > pgbackup-$(date +%F).tar.gz

**Comment by mbosecke** [comment]
>There is no "sync" done when you use backup_script. A temporary folder is created, your command gets executed in this folder and after that, it will get copied with plain cp. rsync is not used for backup_script for this but that is what rsnapreport.pl depends on.

We're on the same page here; I'm suggesting that we replace "plain cp" with rsync during this phase. You call it copy, I call it sync, the source code calls it *sync_if_different* but we're talking about the same thing.

> but you won't get statistics how long the network-transfer needed, and so on.

I agree that that is an issue that would need to be resolved. I would go so far as arguing that rsnapshot should not rely on rsync for the timing-related statistics; all non-rsync overhead is completely ignored by such a method (which I admit will often be negligible but not necessarily). 

I'm a little surprised there's resistance here. The `backup_script` directive has no dependency on the `backup` directive, it generates content in the backup root, therefore it can be used as a standalone backup strategy and yet it generates no usable statistics. If I wasn't using `backup_script` to actually back up data, I would use `cmd_preexec` instead which DOES require a seperate `backup` directive to be used.

> In your case, I would just use a normal shell-script which is executed by a cronjob and looks a bit like this.

My use case is actually a lot more complex; I only simplified it to be used as an example for the sake of this discussion.
